promql/generated_parser.y

```diff
@@ -35,120 +36,351 @@ import (
+                        { $$ = $2 }
+                | LEFT_PAREN grouping_label_list COMMA RIGHT_PAREN
+                        { $$ = $2 }
+                | LEFT_PAREN RIGHT_PAREN
+                        { $$ = []string{} }
+                | error
+                        { yylex.(*parser).unexpected("grouping opts", "\"(\"") }
+                ;
+
+
+grouping_label_list:
+                grouping_label_list COMMA grouping_label
+                        { $$ = append($1, $3.Val) }
+                | grouping_label
+                        { $$ = []string{$1.Val} }
+                | grouping_label_list error
+                        { yylex.(*parser).unexpected("grouping opts", "\",\" or \")\"") }
+                ;
+
+grouping_label  : maybe_label
+                        {
+                        if !isLabel($1.Val) {
+                                yylex.(*parser).unexpected("grouping opts", "label")
+                        }
+                        $$ = $1
+                        }
+                | error
+                        { yylex.(*parser).unexpected("grouping opts", "label") }
+                ;
+
+/*
+ * Function calls.
+ */
+                
+function_call   : IDENTIFIER function_call_body
+                        {
+                        fn, exist := getFunction($1.Val)
+                        if !exist{
+                                yylex.(*parser).errorf("unknown function with name %q", $1.Val)
+                        } 
+                        $$ = &Call{
+                                Func: fn,
+                                Args: $2.(Expressions),
+                        }
+                        }
+                ;
+
+function_call_body: LEFT_PAREN function_call_args RIGHT_PAREN
+                        { $$ = $2 }
+                | LEFT_PAREN RIGHT_PAREN
+                        {$$ = Expressions{}}
+                ;
+
+function_call_args: function_call_args COMMA expr
+                        { $$ = append($1.(Expressions), $3.(Expr)) }
+                | expr 
+                        { $$ = Expressions{$1.(Expr)} }
+                ;
+
+/*
+ * Expressions inside parentheses.
+ */
+
+paren_expr      : LEFT_PAREN expr RIGHT_PAREN
+                        { $$ = &ParenExpr{Expr: $2.(Expr)} }
+                ;
+
+/*
+ * Offset modifiers.
+ */
+
+offset_expr: expr OFFSET duration
+                        {
+                        yylex.(*parser).addOffset($1, $3)
+                        $$ = $1
+                        }
+                | expr OFFSET error
+                        { yylex.(*parser).unexpected("offset", "duration") }
+                ;
+
+/*
+ * Subquery and range selectors.
+ */
+
+matrix_selector : expr LEFT_BRACKET duration RIGHT_BRACKET
+                        {
+                        vs, ok := $1.(*VectorSelector)
+                        if !ok{
+                                yylex.(*parser).errorf("matrix selectors only allowed for vector selectors")
```

Comment:
Avoid the use of the word matrix in anything user facing, it's archaic in Prometheus. This is a range on a selector.

Reply:
Updated this.